### PR TITLE
Corgi/quit room

### PR DESCRIPTION
**Pull request by tillywoodfield**

### CODE REVIEW

File: server/controllers/index.js [hunk 1]
```diff
@@ -52,43 +54,65 @@ router.get('/auth/logout', auth, (req, res) => {
...
-        // Add the creator to the room
-        if (newGameState.canJoinGame(user)) { /* Game has not been started yet */
-            newGameState.joinGame(user, () => {
-                req.session.roomId = roomId;    
-                res.sendStatus(200);
-                roomId++;
-            });
-        } else { /* Game has already began */
-            res.status(400).json({message: "Game has already started."});
+        if (req.session.roomId !== undefined) {
```
Comment (sibela46) [inline server/controllers/index.js]: if (!req.session.roomId) ?

File: cypress/integration/game_spec.js [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+                cy.voteCard();
+            });
+        });
+
+        it('redirects back to the lobby', () => {
+            cy.newGame();
+            cy.url().should('include', '/lobby');
+        });
+    });
+});
```
Comment (sibela46) [inline cypress/integration/game_spec.js]: new line

File: cypress/integration/lobby_spec.js [hunk 3]
```diff
@@ -1,71 +1,105 @@
...
-                cy.playCard();
-                cy.voteCard();
-                cy.wait(1000);
-                cy.newGame();
-                cy.url().should('include', '/lobby');
-            })
+    describe('on logging in when a room already exists', () => {
+        beforeEach(() => {
+            cy.request(`http://localhost:12346/connect?url=${encodeURIComponent(url)}`).then(() => {
+                cy.request(`http://localhost:12346/createRoom?url=${encodeURIComponent(url)}`).then(() => {
```
Comment (sibela46) [inline cypress/integration/lobby_spec.js]: you could chain these as returns

File: cypress/integration/play_card_spec.js [hunk 4]
```diff
@@ -4,14 +4,18 @@ describe('Play card', () => {
...
-            cy.request(`http://localhost:12346/connect?url=${encodeURIComponent(Cypress.config().baseUrl)}`)
-            .then(() => {
-                cy.startGame();
-                cy.request(`http://localhost:12346/playCardWord?url=${encodeURIComponent(Cypress.config().baseUrl)}`)
-            })
-            .then(() => cy.get('[data-cy="play-card"]').should('exist'));
+            cy.request(`http://localhost:12346/connect?url=${encodeURIComponent(url)}`).then(() => {
+                cy.request(`http://localhost:12346/joinRoom?roomId=0&url=${encodeURIComponent(Cypress.config().baseUrl)}`).then(() => {
+                    cy.startGame();
+                    cy.request(`http://localhost:12346/playCardWord?url=${encodeURIComponent(Cypress.config().baseUrl)}`);
```
Comment (sibela46) [inline cypress/integration/play_card_spec.js]: same as above

File: server/services/socket.js [hunk 5]
```diff
@@ -12,6 +14,7 @@ exports.setupSocket = (server, session) => {
         if (socket.handshake.session.user) {
             sockets = sockets.filter(otherSocket => otherSocket.handshake.session.user !== socket.handshake.session.user);
             sockets.push(socket);
+            emitRooms();
```
Comment (sibela46) [inline server/services/socket.js:17]: does this have to be this.emitRooms()

File: server/controllers/index.js (hunk 1 above)
Comment (tillywoodfield) [reply server/controllers/index.js]: Doesn't work as the first roomId is 0

File: cypress/integration/lobby_spec.js [hunk 6]
```diff
@@ -1,71 +1,105 @@
...
+                cy.get('[data-cy="player-username"]').contains('unicorn');
+                cy.get('[data-cy="player-username"]').contains('halfling');
+            });
+
+            it('displays the start game button', () => {
+                cy.get('[data-cy="start-game"]').should('exist');
+            });
+
+            describe('on clicking start game button', () => {
+                it('redirects to the dahsboard', () => {
```
Comment (Ro4052) [inline cypress/integration/lobby_spec.js]:
```suggestion
                it('redirects to the dashboard', () => {
```

File: cypress/integration/lobby_spec.js (hunk 3 above)
Comment (Ro4052) [inline cypress/integration/lobby_spec.js]: This is another place where we could chain promises.

File: cypress/integration/play_card_spec.js [hunk 7]
```diff
@@ -4,14 +4,18 @@ describe('Play card', () => {
...
+        beforeEach(() => {
             cy.createRoom();
-            cy.request(`http://localhost:12346/connect?url=${encodeURIComponent(Cypress.config().baseUrl)}`)
-            .then(() => {
-                cy.startGame();
-                cy.request(`http://localhost:12346/playCardWord?url=${encodeURIComponent(Cypress.config().baseUrl)}`)
-            })
-            .then(() => cy.get('[data-cy="play-card"]').should('exist'));
+            cy.request(`http://localhost:12346/connect?url=${encodeURIComponent(url)}`).then(() => {
+                cy.request(`http://localhost:12346/joinRoom?roomId=0&url=${encodeURIComponent(Cypress.config().baseUrl)}`).then(() => {
```
Comment (Ro4052) [inline cypress/integration/play_card_spec.js]: And here.

File: cypress/integration/play_word_and_card_spec.js [hunk 8]
```diff
@@ -1,25 +1,47 @@
...
-        cy.startGame();
-    });
+const url = Cypress.config().baseUrl;
 
+describe('Play word and card', () => {
     describe('on start game', () => {
+        beforeEach(() => {
+            cy.login('unicorn');
+            cy.request(`http://localhost:12346/connect?url=${encodeURIComponent(url)}`).then(() => {
+                cy.request(`http://localhost:12346/createRoom?url=${encodeURIComponent(Cypress.config().baseUrl)}`).then(() => {
```
Comment (Ro4052) [inline cypress/integration/play_word_and_card_spec.js]: Could chain here as well. It's really a stylistic decision, whatever you think is more readable.

File: server/controllers/index.js [hunk 9]
```diff
@@ -52,43 +54,65 @@ router.get('/auth/logout', auth, (req, res) => {
...
+            const oldRoomId = req.session.roomId;
+            let game = getGameStateById(oldRoomId);
```
Comment (Ro4052) [inline server/controllers/index.js]: Can this be a `const`?
```suggestion
            const game = getGameStateById(oldRoomId);
```

File: server/controllers/index.js [hunk 10]
```diff
@@ -52,43 +54,65 @@ router.get('/auth/logout', auth, (req, res) => {
...
+            game.quitGame(user);
+            if (!game.getPlayers().length) games = games.filter(otherGame => otherGame !== game);
+            leaveRoom(user, oldRoomId);
+            req.session.roomId = undefined;
         }
+        let newGameState = new GameLogic(latestRoomId);
```
Comment (Ro4052) [inline server/controllers/index.js]:
```suggestion
        const newGameState = new GameLogic(latestRoomId);
```

File: server/controllers/index.js [hunk 11]
```diff
@@ -52,43 +54,65 @@ router.get('/auth/logout', auth, (req, res) => {
...
+        let newGameRoom = {roomId: latestRoomId, gameState: newGameState}
```
Comment (Ro4052) [inline server/controllers/index.js]:
```suggestion
        const newGameRoom = {roomId: latestRoomId, gameState: newGameState}
```

File: server/controllers/index.js [hunk 12]
```diff
@@ -52,43 +54,65 @@ router.get('/auth/logout', auth, (req, res) => {
...
-        if (getGameStateById(roomId).canJoinGame(user)) {
-            getGameStateById(roomId).joinGame(user, () => {
-                req.session.roomId = roomId; 
-                res.sendStatus(200);
-            });
-        } else { /* Game has already began */
-            res.status(400).json({message: "Game has already started."});
+        if (req.session.roomId !== undefined) {
+            const oldRoomId = req.session.roomId;
+            let game = getGameStateById(oldRoomId);
```
Comment (Ro4052) [inline server/controllers/index.js]:
```suggestion
            const game = getGameStateById(oldRoomId);
```

File: server/controllers/index.js [hunk 13]
```diff
@@ -52,43 +54,65 @@ router.get('/auth/logout', auth, (req, res) => {
...
+            game.quitGame(user);
+            if (!game.getPlayers().length) games = games.filter(otherGame => otherGame !== game);
+            leaveRoom(user, oldRoomId);
+            req.session.roomId = undefined;
```
Comment (Ro4052) [inline server/controllers/index.js]: I don't usually like setting something to `undefined` myself. I'd prefer to use `null`. This is depends on you changing the `req.session.roomId !== undefined` line above.
```suggestion
            req.session.roomId = null;
```

File: server/controllers/index.js [hunk 14]
```diff
@@ -52,43 +54,65 @@ router.get('/auth/logout', auth, (req, res) => {
...
+        res.status(400).json({message: err.message})
+    }
+});
+
+/* Leave a room */
+router.post('/api/room/leave', auth, (req, res) => {
+    const { user } = req.session;
+    const { roomId } = req.body;
+    try {
+        let game = getGameStateById(roomId);
```
Comment (Ro4052) [inline server/controllers/index.js]:
```suggestion
        const game = getGameStateById(roomId);
```

File: server/services/GameLogic.js [hunk 15]
```diff
@@ -6,7 +6,7 @@ const { statusTypes } = require('./statusTypes');
 const promptDuration = process.env.NODE_ENV === 'testing' ? 1000 : 37000;
 const nextRoundDuration = process.env.NODE_ENV === 'testing' ? 1000 : 5000;
 const rounds = 3;
-const minPlayers = 0;
+const minPlayers = 2;
```
Comment (Ro4052) [inline server/services/GameLogic.js]: Does the minimum number of players not need to be 3?

File: server/services/GameLogic.js [hunk 16]
```diff
@@ -34,6 +34,9 @@ module.exports = class GameLogic {
         socket.emitPlayers(this.roomId, this.players);
     }
 
+    /* Get the minimum number of players */
+    getMinPlayers() { return minPlayers };
```
Comment (Ro4052) [inline server/services/GameLogic.js]: Why do you need this method. Can't you just use the global variable?

File: server/services/socket.js [hunk 17]
```diff
@@ -34,18 +37,36 @@ exports.createRoom = (username, roomId) => {
 
 // Join an existing room
 exports.joinRoom = (username, roomId) => {
-    let room = rooms.find((room) => room.id === roomId);
-    if (username !== room.creator && !room.players.some((player) => player === username)) {
+    let room = rooms.find((room) => room.roomId === roomId);
```
Comment (Ro4052) [inline server/services/socket.js]:
```suggestion
    const room = rooms.find((room) => room.roomId === roomId);
```

File: server/services/socket.js [hunk 18]
```diff
@@ -34,18 +37,36 @@ exports.createRoom = (username, roomId) => {
...
+        room.players.push({ username });
         this.emitRooms();
     } else {
         throw new Error("You cannot join the room again.");
     }
 };
 
+// Leave a room
+exports.leaveRoom = (username, roomId) => {
+    let room = rooms.find((room) => room.roomId === roomId);
```
Comment (Ro4052) [inline server/services/socket.js]:
```suggestion
    const room = rooms.find((room) => room.roomId === roomId);
```

File: server/services/socket.js [hunk 19]
```diff
@@ -34,18 +37,36 @@ exports.createRoom = (username, roomId) => {
...
+    let socket = sockets.find(socket => socket.handshake.session.user === username);
```
Comment (Ro4052) [inline server/services/socket.js]:
```suggestion
    const socket = sockets.find(socket => socket.handshake.session.user === username);
```

File: server/services/socket.js [hunk 20]
```diff
@@ -34,18 +37,36 @@ exports.createRoom = (username, roomId) => {
...
+    socket.leave(`room-${roomId}`);
+    socket.handshake.session.roomId = undefined;
+    room.players = room.players.filter(player => player.username !== username );
+    if (room.players.length <= 0) rooms = rooms.filter(otherRoom => otherRoom !== room);
+    this.emitRooms();
+};
+
+// Set that the room has started
+exports.setRoomStarted = (roomId) => {
+    let room = rooms.find(room => room.roomId === roomId);
```
Comment (Ro4052) [inline server/services/socket.js]:
```suggestion
    const room = rooms.find(room => room.roomId === roomId);
```

**Review by Ro4052 — COMMENTED** [review]
Looks like more than there really is. Just a few things that happen a number of times.

**Review by sibela46 — APPROVED** [review]

**Pull request merged**